### PR TITLE
Define KnockoutObservableType to match returned type of fromJS

### PR DESCRIPTION
**Pull request by turdwaster**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: Uhm, https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html ?
- [X] Increase the version number in the header if appropriate. - there was none - changed TS version
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by turdwaster** [comment]
`Error: knockout depends on knockout.mapping but has a lower required TypeScript version.` - does knockout depend on knockout.mapping or am I misinterpreting this? I would expect it to be the other way around. Please advise on how to handle TS 2.8 features in this case...

File: types/knockout/index.d.ts [hunk 1]
```diff
@@ -9,7 +9,6 @@
 //                 Leonardo Lombardi <https://github.com/ltlombardi>
 //                 Retsam <https://github.com/Retsam>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
```
Comment (Igorbek) [inline types/knockout/index.d.ts]: was removing TS version required?

File: types/knockout.mapping/index.d.ts [hunk 2]
```diff
@@ -13,8 +14,14 @@ declare var self: KnockoutMapping;
 export = self;
 
 declare global {
-    type KnockoutObservableType<T> = {	
-        [P in keyof T]: KnockoutObservable<KnockoutObservableType<T[P]>>|T[P];	
+	type KnockoutObservableType<T> = {
```
Comment (Igorbek) [inline types/knockout.mapping/index.d.ts]: I'm not really familiar with `knockout.mapping`, but by looking at type definition I can suggest
- extract conditional type to another helper type
- test for primitive types first without going into mapped type

something like:

```ts
type KnockoutObservableType<T> =
  T extends Date | string | boolean | number ? KnockoutObservable<T> :
  T extends (infer E)[] ? KnockoutObservableArray<KnockoutObservableType<E>> :
  T extends object ? { [P in keyof T]: KnockoutObservableType<T[P]> } :
  unknown;
```

File: types/knockout/index.d.ts (hunk 1 above)
Comment (turdwaster) [reply types/knockout/index.d.ts]: This is a feeble attempt to make it build I still have the problem of https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32210#issuecomment-454794822 that I don't know how to solve :-(

File: types/knockout.mapping/index.d.ts (hunk 2 above)
Comment (turdwaster) [reply types/knockout.mapping/index.d.ts]: That is really pretty, but I get `error TS2456: Build:Type alias 'KnockoutObservableType' circularly references itself.` from it. Taking the good part of the comment above, I end up with
```typescript
type KnockoutObservableType<T> = {
    [P in keyof T]:
    T[P] extends Date | string | boolean | number ? KnockoutObservable<T[P]> :
    T[P] extends (infer E)[] ? KnockoutObservableArray<KnockoutObservableType<E>> :
    T[P] extends object ? T[P] :
    unknown;
};
```

Better?

**Pull request closed without merging**